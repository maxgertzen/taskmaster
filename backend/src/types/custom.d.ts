import { Express, Request } from "express-serve-static-core";

declare module "express" {
  export interface Request {
    userId?: string;
  }
}
