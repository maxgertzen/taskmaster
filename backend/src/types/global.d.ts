declare namespace NodeJS {
  interface ProcessEnv {
    PORT?: string;
    REDIS_HOST: string;
    REDIS_PORT: string;
    REDIS_PASSWORD: string;
    LOCAL_DEV_CLIENT_URL: string;
    NODE_ENV: "development" | "production" | "test";
  }
}
