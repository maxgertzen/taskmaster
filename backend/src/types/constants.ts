export type DBType = "redis" | "mongo";
