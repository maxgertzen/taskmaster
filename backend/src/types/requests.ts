import { Request } from "express";

export type CreateListRequestBody = {
  name: string;
};

export interface CreateListRequest extends Request {
  body: CreateListRequestBody;
}

export type UpdateListRequestBody = {
  id: string;
  name: string;
};

export interface UpdateListRequest extends Request {
  body: UpdateListRequestBody;
}

export type DeleteRequestBody = {
  id: string;
};

export interface DeleteListRequest extends Request {
  body: DeleteRequestBody;
}

export type ReorderListRequestBody = {
  oldIndex: number;
  newIndex: number;
};

export interface ReorderListRequest extends Request {
  body: ReorderListRequestBody;
}

export type CreateTaskRequestBody = {
  text: string;
  listId: string;
};

export interface CreateTaskRequest extends Request {
  body: CreateTaskRequestBody;
}

export type GetTasksRequestParams = {
  listId: string;
};

export interface GetTasksRequest extends Request {
  params: GetTasksRequestParams;
}

export type UpdateTaskRequestBody = {
  id: string;
  text?: string;
  completed?: boolean;
};

export interface UpdateTaskRequest extends Request {
  body: UpdateTaskRequestBody;
}

export type DeleteTaskRequestBody = {
  taskId: string;
  listId: string;
};

export interface DeleteTaskRequest extends Request {
  body: DeleteTaskRequestBody;
}

export type ReorderTasksRequestBody = {
  listId: string;
  oldIndex: number;
  newIndex: number;
};

export interface ReorderTasksRequest extends Request {
  body: ReorderTasksRequestBody;
}

export type BulkCompleteRequestBody = {
  listId: string;
  newCompletedState: boolean;
};
export interface BulkCompleteRequest extends Request {
  body: BulkCompleteRequestBody;
}

export type BulkDeleteRequestBody = {
  listId: string;
  mode?: "all" | "completed";
};

export interface BulkDeleteRequest extends Request {
  body: BulkDeleteRequestBody;
}
