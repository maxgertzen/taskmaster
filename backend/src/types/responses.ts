import {
  List,
  ClientTask,
  SearchResults,
  BaseUser,
} from "../interfaces/entities";

export type CreateListResponse = List;

export type GetListsResponse = List[];

export type UpdateListResponse = List;

export type DeleteListResponse = {
  deletedId: string;
};

export type ReorderListsResponse = List[];

export type CreateTaskResponse = ClientTask;

export type GetTasksResponse = ClientTask[];

export type GetTasksSearchResultsResponse = SearchResults;

export type UpdateTaskResponse = string;

export type DeleteTaskResponse = {
  deletedId: string;
};

export type ReorderTasksResponse = ClientTask[];

export type ToggleCompleteAllResponse = ClientTask[];

export type BulkDeleteResponse = ClientTask[];

export type GetOrCreateUserResponse = BaseUser;

export type UpdatePreferencesResponse = BaseUser | null;
