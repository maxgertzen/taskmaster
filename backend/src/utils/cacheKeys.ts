export const CACHE_KEYS = {
  TASK: (userId: string, taskId: string) =>
    `cache:user:${userId}:task:${taskId}` as const,
  TASK_LIST: (userId: string, listId: string) =>
    `cache:user:${userId}:tasks:${listId}` as const,
  LIST: (userId: string, listId: string) =>
    `cache:user:${userId}:list:${listId}` as const,
  LISTS: (userId: string) => `cache:user:${userId}:lists` as const,
  SEARCH_RESULTS: (userId: string, query: string) =>
    `cache:user:${userId}:search:${query}` as const,
} as const;
