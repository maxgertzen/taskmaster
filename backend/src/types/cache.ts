import { CACHE_KEYS } from "../utils/cacheKeys";

export type CacheConfig = {
  ttl: number;
  enabled: boolean;
};

type CacheKeyFunction = (typeof CACHE_KEYS)[keyof typeof CACHE_KEYS];
export type CacheKey = ReturnType<CacheKeyFunction>;
