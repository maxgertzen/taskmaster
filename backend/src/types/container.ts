import { makeListsController } from "@src/controllers/listController";
import { makeTasksController } from "@src/controllers/taskController";
import { IListRepository } from "@src/interfaces/listRepository";
import { ITaskRepository } from "@src/interfaces/taskRepository";
import { IUserRepository } from "@src/interfaces/userRepository";
import { ListsService, TasksService, UsersService } from "@src/services";
import { ListsCache, TasksCache, UsersCache } from "@src/services/cache";
import { CacheService } from "@src/services/cache/cacheService";
import { RedisClientType } from "redis";

export interface ContainerType {
  redisClient: RedisClientType;
  cacheService: CacheService;
  listsCache: ListsCache;
  tasksCache: TasksCache;
  usersCache: UsersCache;
  listRepository: IListRepository;
  taskRepository: ITaskRepository;
  userRepository: IUserRepository;
  listsController: ReturnType<typeof makeListsController>;
  tasksController: ReturnType<typeof makeTasksController>;
  listsService: ListsService;
  tasksService: TasksService;
  usersService: UsersService;
}
