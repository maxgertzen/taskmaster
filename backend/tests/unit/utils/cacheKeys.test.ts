import { CACHE_KEYS } from "@src/utils/cacheKeys";
import { CacheKey } from "@tests/types/utils";
import { generateExpectedKey } from "@tests/helpers/cacheKeys";
import { listFactory, taskFactory, userFactory } from "@tests/data";

describe("CACHE_KEYS", () => {
  const user = userFactory.generateBaseUser();
  const list = listFactory.generateBaseList();
  const task = taskFactory.generateBaseTask();

  const cacheKeys = Object.keys(CACHE_KEYS) as CacheKey[];

  describe("valid inputs", () => {
    const keyToArgs: Record<CacheKey, string[]> = {
      TASK: [task.userId, task.id],
      TASK_LIST: [list.userId, list.id],
      LIST: [list.userId, list.id],
      LISTS: [list.userId],
      SEARCH_RESULTS: [user.auth0Id, "query"],
      USER: [user.auth0Id],
    };

    cacheKeys.forEach((key) => {
      it(`should generate correct cache key for ${key}`, () => {
        const args = keyToArgs[key];
        const fn = CACHE_KEYS[key];
        const expected = generateExpectedKey(key, args);

        expect(fn(...(args as [string, string]))).toBe(expected);
      });
    });
  });

  describe("edge cases", () => {
    it("should handle empty userId", () => {
      const result = CACHE_KEYS.TASK("", task.id);
      const expected = generateExpectedKey("TASK", ["", task.id]);
      expect(result).toBe(expected);
    });
  });
});
