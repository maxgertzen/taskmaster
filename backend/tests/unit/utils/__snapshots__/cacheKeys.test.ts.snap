// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CACHE_KEYS should match the expected structure snapshot 1`] = `
{
  "LIST": [Function],
  "LISTS": [Function],
  "SEARCH_RESULTS": [Function],
  "TASK": [Function],
  "TASK_LIST": [Function],
  "USER": [Function],
}
`;
